plugins/woocommerce/tests/php/includes/class-wc-shortcodes-test.php

```diff
@@ -476,4 +476,30 @@ public function test_product_page_shortcode_protected_product() {
 
 		$this->assertStringContainsString( 'This content is password protected', $product_page );
 	}
+
+	/**
+	 * Ensure that when both [woocommerce_cart] and [woocommerce_checkout] shortcodes
+	 * are rendered together, no redirect to the cart page occurs.
+	 */
+	public function test_combined_cart_and_checkout_shortcodes_do_not_redirect() {
+		// Simulate being on the cart page (as if [woocommerce_cart] is present).
+		add_filter( 'woocommerce_is_cart', '__return_true' );
+
+		// Track if a redirect is triggered.
+		$redirect_url = null;
+		add_filter( 'wp_redirect', function ( $location ) use ( &$redirect_url ) {
+			$redirect_url = $location;
+			return $location;
+		}, 10 );
```

Comment:
Lets refactor this slightly so we can clean up this filter after the test has run:

* Move `$redirect_url` to a variable on the class.
* Move the anonymous function into a helper method on this class also.

I imagine it looking something like this:

```php
/**
* Helper method to track redirects.
*
* @param string $location The redirect location.
* @return string The unchanged redirect location.
*/
public function track_redirect( $location ) {
	$this->redirect_url = $location;
	return $location;
}
```

Then we can reference it when adding the filter and also removing it:

```php
add_filter( 'wp_redirect', array( $this, 'track_redirect' ), 10, 1 );
remove_filter( 'wp_redirect', array( $this, 'track_redirect' ), 10 );
```